libsolidity/ast/Types.cpp

```diff
@@ -2252,6 +2266,41 @@ u256 StructType::storageSize() const
 	return max<u256>(1, members(nullptr).storageSize());
 }
 
+bool StructType::containsNestedMapping() const
+{
+	if (!m_struct.annotation().containsNestedMapping.has_value())
+	{
+		struct WrappedType
+		{
+			TypePointer typePointer;
+			WrappedType(TypePointer type): typePointer(type) { solAssert(typePointer, ""); }
+			bool operator<(WrappedType const& other) const
+			{
+				return typePointer->richIdentifier() < other.typePointer->richIdentifier();
```

Comment:
I don't have a good feeling about this either, sorry :)
Maybe you can base the decision about whether or not to stop visiting on the struct declarations? It maybe does not matter if we handle the same integer type multiple times - maybe you can even add a shortcut that checks if it is a reference type or a mapping, anything else we do not care about.

Reply:
> Maybe you can base the decision about whether or not to stop visiting on the struct declarations? 

I don't quite understand. Are we still talking about doing a BFS on `Type const*`? If yes, I could skip adding elements other than struct, arrays, or mappings. This may not look very clean though.

Also, can't we use `richIdentifier()` to compare types?

Reply:
We have no real guarantee that `richIdentifier()` and `Type::operator==` are compatible.

My suggestion was to do a different BFS on ASTNodes instead.